Feature/listener refactor

jagill opened the pull request:
@jjackson26 @swarajban This is refactor of the `recognizer.onresult` callback, which simplifies it a bit, as well as allowed me to find a bug.

The gist of it is:
1. Separate out the logic that parses the recognizer result event into a listener result event.
2. Instead of storing an interim result as the final entry of the `listener._results` object, it stores it in its own `listener._pendingResult` variable, which makes the logic for handling it much more comprehensible.

There are also some additions that allow recording events, as well as a FakeRecognizer helper class for testing, and new tests.  But these are secondary.

jjackson26 commented on src/sdk/listener.js:
```diff
@@ -2,6 +2,36 @@
+      pad(date.getMilliseconds(), 3) +
+    ']';
+  };
+  // ZZZ END DEBUG STUFF
```
Do we want to remove this debug code?

jjackson26 commented on src/sdk/listener.js:
```diff
@@ -415,33 +449,35 @@
-          lastResult.early = true;
-          listener.emit('result', lastResult, resultIndex, results, event);
-        }
+      if (listener._pendingResult) {
```
Style: maybe change `function finalizeResult() {` to `var finalizeResult = function () {` for consistency.

jjackson26 commented on src/sdk/listener.js:
```diff
@@ -496,22 +505,63 @@
+      // because they are duplicates) but when we get final results we
+      // want to start listening again.  This, when the result is final
+      // set resultFinalzed = false.
+      if (pendingResultFinalized && result.final) {
```
Nice clear explanation.

swarajban replied:
+1

jjackson26 commented on src/test/spec/FakeRecognizer.js:
```diff
@@ -0,0 +1,118 @@
+    jasmine.clock().uninstall();
+  };
+
+})();
```
Cool stuff

swarajban replied:
Ya, cool man

jjackson26 commented:
LGTM. It would be nice to have Swaraj see if he can easily make his way through the logic, as he's less acquainted with it.

swarajban commented on src/sdk/listener.js:
```diff
@@ -496,22 +505,63 @@
 
       // if we restarted, we'll need to add a space for some results
-      if (resultIndex >= 0 && !/^\s/.test(result.transcript.charAt(0))) {
+      if (listener._results.length > 0 && !/^\s/.test(result.transcript.charAt(0))) {
```
not relevant to this PR really, but why is the whitespace test a regex? Why can't it be `result.transcript.charAt(0) !== ' '`

jagill replied:
Good question -- I'm still a little unsure about when we need this;  I'd love some more tests showing the cases.

swarajban commented on src/sdk/listener.js:
```diff
@@ -496,22 +505,63 @@
+        listener._results.push(result);
+      } else {
+        // Remember this in case we need to finalize it.
+        listener._pendingResult = result;
```
I guess I forgot, but we only store one _pendingResult?

jagill replied:
The flow is that we get N interimResults, usually followed by a a final result (that's the completion of those interim results).  The interim results are like ['hi', 'hi how', 'hi, how are', 'hi, how are you'], so we only want the last (which has the highest confidence).  Thus, at any given time, there is only one relevant interim result, which is the latest.

swarajban commented on src/sdk/listener.js:
```diff
@@ -415,33 +449,35 @@
-      if (listener._captureRaw) {
-        listener._rawResults.push(event);
-      }
+    var registerEvent = function (name, e) {
```
Can you add a comment describing what this function does?

swarajban replied:
Ah, just for unit testing ya?

swarajban commented on src/sdk/main.js:
```diff
@@ -859,7 +859,6 @@ var MM = ( function (window, ajax, Faye) {
               MM.activeUser.sessions.post(
                 session,
                 function onSessionCreate(result) {
-                  console.log('Create session results:', result);
```
Cool, are there any more of these lying around?

jagill replied:
Not that I saw.

swarajban commented on src/test/spec/FakeRecognizer.js:
```diff
@@ -0,0 +1,118 @@
+
+;(function () {
+
+  var FakeRecognizer = window.FakeRecognizer = function () {
```
Can you add a brief description of this class in comments (I know *we* all know it's a fake webkit recognizer, but let's write it down here)

swarajban commented on src/test/spec/FakeRecognizer.js:
```diff
@@ -0,0 +1,118 @@
+
+  FakeRecognizer.prototype.start = function () {
+    if (this._isStarting) {
+      // var e = new DOMError("InvalidStateError");
```
what is this?

swarajban commented on src/test/spec/FakeRecognizer.js:
```diff
@@ -0,0 +1,118 @@
+
+  FakeRecognizer.prototype._end = function () {
+    this._isStarting = false;
+    this.onend && this.onend();
```
Just supports a single event subscriber ya?

jagill replied:
Yes, the (real) recognizer just has `onend` etc hooks that you set; it's unfortunately not evented.

swarajban commented on src/test/spec/FakeRecognizer.js:
```diff
@@ -0,0 +1,118 @@
+  };
+
+  FakeRecognizer.prototype._currentTimestamp = function () {
+    if (this._eventIndex >= this._events.length) {
```
nit: can't you just use a `this._currentEvent === null` check?

swarajban commented on src/test/spec/ListenerPlaybackTest.js:
```diff
@@ -0,0 +1,262 @@
+      listener = new MM.Listener({interimResults: true});
+      listener._recognizer = recognizer;
+      listener._initializeRecognizer(recognizer);
+      recognizer._setEvents(Playbacks.pizza1);
```
haha what a name

jagill replied:
Yeah, i originally was going to name transcripts by the content, but then i started naming them by their function.  I didn't go back to correct the old ones, dunno if I should!

swarajban commented:
This looks a lot better than before. Really like the comments & the recording functionality

jagill commented:
Ok, fixed code review comments.  I've merged a squashed version (there was a large file I added then deleted) and will push that.  Don't merge this PR; i'll close it in a bit.
